lib/autoupdate/installer.go

```diff
@@ -0,0 +1,302 @@
+	}
+	// Check integrity before decompression
+	if !bytes.Equal(newSum, pathSum) {
+		return trace.Errorf("mismatched checksum, download possibly corrupt")
+	}
+	// Get uncompressed size of the tgz
+	n, err := uncompressedSize(f)
+	if err != nil {
+		return trace.Errorf("failed to determine uncompressed size: %w", err)
+	}
+	// Seek to start of tgz after reading size
+	if _, err := f.Seek(0, io.SeekStart); err != nil {
+		return trace.Errorf("failed seek to start: %w", err)
+	}
+	if err := ai.extract(ctx, versionDir, f, n); err != nil {
+		return trace.Errorf("failed to extract teleport: %w", err)
+	}
+	// Write the checksum last. This marks the version directory as valid.
+	err = os.WriteFile(sumPath, []byte(hex.EncodeToString(newSum)), 0755)
+	if err != nil {
+		return trace.Errorf("failed to write checksum: %w", err)
+	}
+	return nil
+}
+
+// makeURL to download the Teleport tgz.
+func makeURL(uriTmpl, version string) (string, error) {
+	tmpl, err := template.New("uri").Parse(uriTmpl)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	var uriBuf bytes.Buffer
+	params := struct {
+		OS, Version, Arch string
+	}{runtime.GOOS, version, runtime.GOARCH}
+	err = tmpl.Execute(&uriBuf, params)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return uriBuf.String(), nil
+}
+
+// readChecksum from the version directory.
+func readChecksum(path string) ([]byte, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer f.Close()
+	var buf bytes.Buffer
+	_, err = io.CopyN(&buf, f, checksumHexLen)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	raw := buf.String()
+	sum, err := hex.DecodeString(raw)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return sum, nil
+}
+
+func (ai *LocalAgentInstaller) getChecksum(ctx context.Context, url string) ([]byte, error) {
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	resp, err := ai.HTTP.Do(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, trace.Errorf("unexpected HTTP status code: %d", resp.StatusCode)
+	}
+
+	// Only attempt to read first 64 bytes
+	var buf bytes.Buffer
+	_, err = io.CopyN(&buf, resp.Body, 64)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	sum, err := hex.DecodeString(buf.String())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return sum, nil
+}
+
+func (ai *LocalAgentInstaller) download(ctx context.Context, w io.Writer, max int64, url string) (sum []byte, err error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	resp, err := ai.HTTP.Do(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
```

Comment:
```golang
	if resp.StatusCode == http.StatusNotFound {
		return "", trace.NotFound("package is not found: %v", url)
	}
```
nit: might be useful for error handling of `Install` method

Reply:
Thanks, adding a check for that status, but not going to use trace.NotFound until I need a typed error.